# **Marketing Analytics**

**PRABU JEYABALAN**

**Part I: Logistic Regression Model:**

In [ ]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.metrics import recall_score, precision_score, f1_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import roc_curve, auc
import matplotlib.pyplot as plt
from sklearn import metrics
import seaborn as sns

**A. Bring the dataset IBM.csv into your environment, and use the head() function to
explore the variables.**

In [ ]:
df = pd.read_csv(r'C:\Users\prabu\Downloads\IBM (1).csv')
df.head()

Age Attrition              Department  DistanceFromHome  Education  \
0   41       Yes                   Sales                 1          2   
1   49        No  Research & Development                 8          1   
2   37       Yes  Research & Development                 2          2   
3   33        No  Research & Development                 3          4   
4   27        No  Research & Development                 2          1   

  EducationField  EnvironmentSatisfaction  JobSatisfaction MaritalStatus  \
0  Life Sciences                        2                4        Single   
1  Life Sciences                        3                2       Married   
2          Other                        4                3        Single   
3  Life Sciences                        4                3       Married   
4        Medical                        1                2       Married   

   MonthlyIncome  NumCompaniesWorked  WorkLifeBalance  YearsAtCompany  
0           5993                   8                1               6  
1           5130                   1                3              10  
2           2090                   6                3               0  
3           2909                   1                3               8  
4           3468                   9                3               2

In [ ]:
df.describe()

Age  DistanceFromHome    Education  EnvironmentSatisfaction  \
count  1470.000000       1470.000000  1470.000000              1470.000000   
mean     36.923810          9.192517     2.912925                 2.721769   
std       9.135373          8.106864     1.024165                 1.093082   
min      18.000000          1.000000     1.000000                 1.000000   
25%      30.000000          2.000000     2.000000                 2.000000   
50%      36.000000          7.000000     3.000000                 3.000000   
75%      43.000000         14.000000     4.000000                 4.000000   
max      60.000000         29.000000     5.000000                 4.000000   

       JobSatisfaction  MonthlyIncome  NumCompaniesWorked  WorkLifeBalance  \
count      1470.000000    1470.000000         1470.000000      1470.000000   
mean          2.728571    6502.931293            2.693197         2.761224   
std           1.102846    4707.956783            2.498009         0.706476   
min           1.000000    1009.000000            0.000000         1.000000   
25%           2.000000    2911.000000            1.000000         2.000000   
50%           3.000000    4919.000000            2.000000         3.000000   
75%           4.000000    8379.000000            4.000000         3.000000   
max           4.000000   19999.000000            9.000000         4.000000   

       YearsAtCompany  
count     1470.000000  
mean         7.008163  
std          6.126525  
min          0.000000  
25%          3.000000  
50%          5.000000  
75%          9.000000  
max         40.000000

**B. Which of the variables here are categorical? Which are numerical? (Note: For
ordinal variables, you can make the call on this – there is not a single ‘correct’
answer for the way to handle them).**

**The numerical variables** *are Age, DistanceFromHome, MonthlyIncome, NumCompaniesWorked, and YearsAtCompany.*


**Categorical variables includes**

**Nominal variables** *such as Attrition, Department, EducationField, and MaritalStatus.*

**Ordinal variables** *Education, EnvironmentSatisfaction, JobSatisfaction, and WorkLifeBalance.*










**C. Use the value_counts() function from pandas to learn more about the outcome
variable, ‘Attrition’**

In [ ]:
df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

**a. Describe your findings -- what are the different outcome classes here,
and how common are each of them in the dataset?**

*The outcome variable "Attrition" has two classes: "No" and "Yes".  "No" is the much more common outcome, with 1233 instances (representing employees who have not left the company), while "Yes" represents 237 instances (employees who have left).  This indicates a class imbalance, where the "No" class significantly outnumbers the "Yes" class.*

**b. In machine learning modeling, what is oversampling? (Note: We won’t
actually do it here – just answer the question, though).**

*In machine learning modeling, oversampling is a technique used to address class imbalance in a dataset.*

*When one class (often the minority class) has significantly fewer examples than another (the majority class), models may perform poorly on the minority class due to the imbalance.*

*Oversampling involves increasing the number of examples in the minority class, either by duplicating existing samples or creating synthetic samples (e.g., using methods like SMOTE - Synthetic Minority Over-sampling Technique). This helps balance the dataset, allowing the model to learn better from both classes and improve its performance, especially in classification tasks.*

**D. Right now, the ‘Attrition’ variable’s values are represented by words – but you
should convert these to numbers.**


  **a. While there are many possible methods for this, you are encouraged to
explore the use of a label encoder.**

In [ ]:
# Initializing LabelEncoder
label_encoder = LabelEncoder()

# Fitting and transforming the 'Attrition' column
df['Attrition'] = label_encoder.fit_transform(df['Attrition'])

# First few rows to verify the changes
df.head()

Age  Attrition              Department  DistanceFromHome  Education  \
0   41          1                   Sales                 1          2   
1   49          0  Research & Development                 8          1   
2   37          1  Research & Development                 2          2   
3   33          0  Research & Development                 3          4   
4   27          0  Research & Development                 2          1   

  EducationField  EnvironmentSatisfaction  JobSatisfaction MaritalStatus  \
0  Life Sciences                        2                4        Single   
1  Life Sciences                        3                2       Married   
2          Other                        4                3        Single   
3  Life Sciences                        4                3       Married   
4        Medical                        1                2       Married   

   MonthlyIncome  NumCompaniesWorked  WorkLifeBalance  YearsAtCompany  
0           5993                   8                1               6  
1           5130                   1                3              10  
2           2090                   6                3               0  
3           2909                   1                3               8  
4           3468                   9                3               2

**E. Missing values. Are there any variables in this dataset for which more than half
the values are missing? If so, remove them. (Note: This is not a trick question –
just answer it literally).**

In [ ]:
df.isnull().sum()

Age                        0
Attrition                  0
Department                 0
DistanceFromHome           0
Education                  0
EducationField             0
EnvironmentSatisfaction    0
JobSatisfaction            0
MaritalStatus              0
MonthlyIncome              0
NumCompaniesWorked         0
WorkLifeBalance            0
YearsAtCompany             0
dtype: int64

*There no missing value or NAN in the dataset . So there is nothing required to be removed*

**a. Why might it be dangerous to impute values for variables with a huge %
of missingness?**

*Imputing values for variables with a high percentage of missing data can be dangerous because it will introduce bias or inaccuracies in the data, leading to unreliable model predictions.*

*When a large portion of data is missing, any imputed values (such as the mean, median, or a predicted value) are likely to represent the actual missing data poorly, potentially skewing the dataset's overall distribution and relationships.*

*This can distort the insights and patterns that the model learns, ultimately decreasing the model’s accuracy and generalizability. Additionally, heavy imputation can mask genuine patterns or anomalies that could be meaningful for analysis.*

**F. Examining correlations**

**a. Build a correlation table to examine the correlations among your numeric
independent variables.**

In [ ]:
# Selecting only the numeric independent variables for correlation analysis:
numeric_cols = ["Age", "DistanceFromHome", "MonthlyIncome", "NumCompaniesWorked", "YearsAtCompany"]
correlation_table = df[numeric_cols].corr()

# Correlation table
correlation_table

Age  DistanceFromHome  MonthlyIncome  \
Age                 1.000000         -0.001686       0.497855   
DistanceFromHome   -0.001686          1.000000      -0.017014   
MonthlyIncome       0.497855         -0.017014       1.000000   
NumCompaniesWorked  0.299635         -0.029251       0.149515   
YearsAtCompany      0.311309          0.009508       0.514285   

                    NumCompaniesWorked  YearsAtCompany  
Age                           0.299635        0.311309  
DistanceFromHome             -0.029251        0.009508  
MonthlyIncome                 0.149515        0.514285  
NumCompaniesWorked            1.000000       -0.118421  
YearsAtCompany               -0.118421        1.000000

**i. Are there any correlations here that are so high as to present a
likely problem with multicollinearity? If so, remove one member
of any highly-correlated pair. If not, keep rolling on.**


*In this correlation table, none of the variable pairs have a correlation coefficient that is high enough to indicate a serious problem with multicollinearity.*

*Typically, multicollinearity concerns arise with correlation values above 0.8 or 0.9.*

*The highest correlation here is between MonthlyIncome and YearsAtCompany at 0.514285, which does not reach a level that would suggest significant multicollinearity. Therefore, there is no need to remove any variables based on multicollinearity concerns in this dataset.*

**G. For any variables that need further treatment, dummify them, being sure to drop
one level as you do.**

In [ ]:
# Nominal and ordinal variables
nominal_columns = ['Department', 'EducationField', 'MaritalStatus']
ordinal_columns = {
    'Education': [1, 2, 3, 4, 5],
    'EnvironmentSatisfaction': [1, 2, 3, 4],
    'JobSatisfaction': [1, 2, 3, 4],
    'WorkLifeBalance': [1, 2, 3, 4]
}

# Converting ordinal variables to ordered categorical
for col, order in ordinal_columns.items():
    df[col] = pd.Categorical(df[col], categories=order, ordered=True)

# Creating dummy variables for both nominal and ordinal columns
df_encoded = pd.get_dummies(df, columns=nominal_columns + list(ordinal_columns.keys()), drop_first=True)
df_encoded = df_encoded.astype(int)


# Encoded data
df_encoded.head()

Age  Attrition  DistanceFromHome  MonthlyIncome  NumCompaniesWorked  \
0   41          1                 1           5993                   8   
1   49          0                 8           5130                   1   
2   37          1                 2           2090                   6   
3   33          0                 3           2909                   1   
4   27          0                 2           3468                   9   

   YearsAtCompany  Department_Research & Development  Department_Sales  \
0               6                                  0                 1   
1              10                                  1                 0   
2               0                                  1                 0   
3               8                                  1                 0   
4               2                                  1                 0   

   EducationField_Life Sciences  EducationField_Marketing  ...  Education_5  \
0                             1                         0  ...            0   
1                             1                         0  ...            0   
2                             0                         0  ...            0   
3                             1                         0  ...            0   
4                             0                         0  ...            0   

   EnvironmentSatisfaction_2  EnvironmentSatisfaction_3  \
0                          1                          0   
1                          0                          1   
2                          0                          0   
3                          0                          0   
4                          0                          0   

   EnvironmentSatisfaction_4  JobSatisfaction_2  JobSatisfaction_3  \
0                          0                  0                  0   
1                          0                  1                  0   
2                          1                  0                  1   
3                          1                  0                  1   
4                          0                  1                  0   

   JobSatisfaction_4  WorkLifeBalance_2  WorkLifeBalance_3  WorkLifeBalance_4  
0                  1                  0                  0                  0  
1                  0                  0                  1                  0  
2                  0                  0                  1                  0  
3                  0                  0                  1                  0  
4                  0                  0                  1                  0  

[5 rows x 28 columns]

**H. Create a data partition. For your random_state value, use a number based on
either your work, home, or school address, or just a number that you like (For
example, I live at 200 Market St, I work at 1010 Commonwealth Avenue, and my
lucky number is 80, so I could use either 200, 1010, or 80). Assign 40% of your
rows to your test set, and 60% to your training set.**

In [ ]:
X = df_encoded.drop('Attrition', axis=1)
y = df_encoded['Attrition']

# Splitting data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state= 911)

# Displaying the sizes of the training and test sets
train_size = X_train.shape[0]
test_size = X_test.shape[0]

train_size, test_size

(882, 588)

**I. Compare the mean values of the variables in the dataset after grouping by
Attrition.**


In [ ]:
grouped_means = df_encoded.groupby('Attrition').mean()
grouped_means

Age  DistanceFromHome  MonthlyIncome  NumCompaniesWorked  \
Attrition                                                                   
0          37.561233          8.915653    6832.739659            2.645580   
1          33.607595         10.632911    4787.092827            2.940928   

           YearsAtCompany  Department_Research & Development  \
Attrition                                                      
0                7.369019                           0.671533   
1                5.130802                           0.561181   

           Department_Sales  EducationField_Life Sciences  \
Attrition                                                   
0                  0.287105                      0.419303   
1                  0.388186                      0.375527   

           EducationField_Marketing  EducationField_Medical  ...  Education_5  \
Attrition                                                    ...                
0                          0.100568                0.325223  ...     0.034874   
1                          0.147679                0.265823  ...     0.021097   

           EnvironmentSatisfaction_2  EnvironmentSatisfaction_3  \
Attrition                                                         
0                           0.197891                   0.317113   
1                           0.181435                   0.261603   

           EnvironmentSatisfaction_4  JobSatisfaction_2  JobSatisfaction_3  \
Attrition                                                                    
0                           0.313058           0.189781           0.299270   
1                           0.253165           0.194093           0.308017   

           JobSatisfaction_4  WorkLifeBalance_2  WorkLifeBalance_3  \
Attrition                                                            
0                   0.330089           0.231955           0.621249   
1                   0.219409           0.244726           0.535865   

           WorkLifeBalance_4  
Attrition                     
0                   0.102190  
1                   0.113924  

[2 rows x 27 columns]

**a. From the results you see here, choose any THREE independent variables
from the dataset, and speculate about their likely impact on the result –
do you think this variable will be strongly impactful? Why or why not?**


**YearsAtCompany** *The mean value for YearsAtCompany in the results indicate the typical length of service. Employees with higher values in YearsAtCompany have lower attrition rates because they've invested more time in the company and may have achieved higher positions or job security. On the other hand, employees with lower values (fewer years at the company) might still be exploring options, which could increase their likelihood of leaving. In the results, a moderately high average suggests that retention being linked to longer tenure.*

**JobSatisfaction** *This variable is split into categories, with higher levels indicating better satisfaction.Values in the JobSatisfaction variables show the distribution across different satisfaction levels ( JobSatisfaction_2, JobSatisfaction_3, JobSatisfaction_4). Employees with higher satisfaction levels are likely less prone to attrition, as job satisfaction often reduces turnover. The distribution of these variables shows a mix of satisfaction levels, where higher satisfaction are associated with lower attrition rates.*

**MonthlyIncome** *The MonthlyIncome variable shows the income distribution among employees. Higher income serve as a retention factor, while lower income contribute to higher attrition if employees feel undercompensated. The range in MonthlyIncome values suggests a diverse pay scale, and those on the lower end were more likely to leave in search of better opportunities. This variable’s moderate correlation with YearsAtCompany also hints that longer tenure may come with better income, which contributes to retention.*

*These observations from the results support the idea that YearsAtCompany, JobSatisfaction, and MonthlyIncome could be influential in determining attrition, as they impact employee satisfaction, loyalty, and financial motivation.*

# **Iteration #1**

**J. Build a logistic regression model using statsmodels, with the outcome variable
‘Attrition’. Use the rest of the remaining variables from the dataset as inputs
Remember to use only your training data to build this model.**


In [ ]:
# Adding a constant to the independent variables
X_train_with_constant = sm.add_constant(X_train)

# logistic regression model
model = sm.Logit(y_train, X_train_with_constant).fit()

Optimization terminated successfully.
         Current function value: 0.366459
         Iterations 8


**K. Show the summary of your model with log_reg.summary(). (Note: If you
named your model something else, e.g. mymodel, you can just use
mymodel.summary() here).**


In [ ]:
# model summary
print(model.summary())

                           Logit Regression Results                           
Dep. Variable:              Attrition   No. Observations:                  882
Model:                          Logit   Df Residuals:                      854
Method:                           MLE   Df Model:                           27
Date:                Sun, 03 Nov 2024   Pseudo R-squ.:                  0.1588
Time:                        19:01:27   Log-Likelihood:                -323.22
converged:                       True   LL-Null:                       -384.25
Covariance Type:            nonrobust   LLR p-value:                 4.741e-14
                                        coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
const                                 1.8256      0.968      1.886      0.059      -0.072       3.723
Age                                  -0.0400      0.015     -2

**a. Which of your numeric variables here are showing high p-values?**

The numeric variables with high p-values (indicating they are not statistically significant at common significance levels, such as 0.05) in the logistic regression model are:

**YearsAtCompany:** p-value = 0.388

**MonthlyIncome:** Although significant at 0.025, it's worth noting that the effect size is very small (coef = -7.96e-05), so its practical impact may be minimal.

These high p-values suggest that these variables may not have a strong or statistically significant impact on predicting Attrition in the model.

**b. For your categorical variables, which ones are showing high p-values for
ALL of the levels in the model summary?**

**Department:**

*Department_Research & Development: p-value = 0.537*

*Department_Sales: p-value = 0.351*

**EducationField:**

*EducationField_Life Sciences: p-value = 0.158*

*EducationField_Marketing: p-value = 0.463*

*EducationField_Medical: p-value = 0.142*

*EducationField_Other: p-value = 0.200*

*EducationField_Technical Degree: p-value = 0.158*

**Education:**

*Education_2: p-value = 0.627*

*Education_3: p-value = 0.800*

*Education_4: p-value = 0.453*

*Education_5: p-value = 0.822*

*These high p-values across all levels indicate that these categorical variables (Department, EducationField, and Education) may not have a statistically significant impact on predicting Attrition in this model.*

# **Iteration #2**

**L. Now, build yet another model. Again use statsmodels, and again, use your
training set only. Start with the variables you used in Iteration #1 but drop the
ones you identified in the previous step, for parts (a) and (b).
a. Show the results of this 2nd model with log_reg.summary().**


In [ ]:
# Selecting remaining variables for Iteration #2
X_train_iteration_2 = X_train_with_constant.drop(columns=[
    'YearsAtCompany', 'MonthlyIncome',  # Numeric variables
    'Department_Research & Development', 'Department_Sales',  # Categorical: Department
    'EducationField_Life Sciences', 'EducationField_Marketing', 'EducationField_Medical',
    'EducationField_Other', 'EducationField_Technical Degree',  # Categorical: EducationField
    'Education_2', 'Education_3', 'Education_4', 'Education_5'  # Categorical: Education
])

# Fitting the logistic regression model for Iteration #2
model_iteration_2 = sm.Logit(y_train, X_train_iteration_2).fit()

Optimization terminated successfully.
         Current function value: 0.386036
         Iterations 7


**a. Show the results of this 2nd model with log_reg.summary().**

In [ ]:
print(model_iteration_2.summary())

                           Logit Regression Results                           
Dep. Variable:              Attrition   No. Observations:                  882
Model:                          Logit   Df Residuals:                      867
Method:                           MLE   Df Model:                           14
Date:                Sun, 03 Nov 2024   Pseudo R-squ.:                  0.1139
Time:                        19:01:27   Log-Likelihood:                -340.48
converged:                       True   LL-Null:                       -384.25
Covariance Type:            nonrobust   LLR p-value:                 1.103e-12
                                coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
const                         1.0192      0.694      1.468      0.142      -0.342       2.380
Age                          -0.0684      0.013     -5.253      0.000      -0.094     

In [ ]:
# Initializing and training the logistic regression model
log_reg_sklearn = LogisticRegression(random_state= 911,max_iter= 1000)
log_reg_sklearn.fit(X_train_iteration_2, y_train)


LogisticRegression(max_iter=1000, random_state=911)

**N. Assess the performance of your model against the test set. Build a confusion
matrix, and answer the following questions about your model. You can use
Python functions to answer any of these questions or you can use your confusion
matrix to determine the answers in a slightly more manual way. The ‘positive’
class in this model is represented by the “1” outcome.**

**a. What is your model’s accuracy rate?**

**b. What is your model’s sensitivity rate?**

**c. What is your model’s specificity rate?**

**d. What is your model’s precision?**

**e. What is your model’s balanced accuracy?**

In [ ]:
# Preparing the test data for prediction
X_test_iteration_2 = sm.add_constant(X_test)
X_test_iteration_2 = X_test_iteration_2[X_train_iteration_2.columns]

# Making predictions on the test set
y_pred = log_reg_sklearn.predict(X_test_iteration_2)

# Confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Calculating evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
sensitivity = cm[1, 1] / (cm[1, 0] + cm[1, 1])
specificity = cm[0, 0] / (cm[0, 0] + cm[0, 1])
precision = cm[1, 1] / (cm[1, 1] + cm[0, 1])
balanced_accuracy = (sensitivity + specificity) / 2

print("Confusion Matrix:\n", cm)
print(f"\nAccuracy: {accuracy}")
print(f"Sensitivity: {sensitivity}")
print(f"Specificity: {specificity}")
print(f"Precision: {precision}")
print(f"Balanced Accuracy: {balanced_accuracy}")

Confusion Matrix:
 [[488   2]
 [ 92   6]]

Accuracy: 0.8401360544217688
Sensitivity: 0.061224489795918366
Specificity: 0.9959183673469387
Precision: 0.75
Balanced Accuracy: 0.5285714285714286


**O. Compare your model’s accuracy against the training set vs. accuracy against the
test set (just use accuracy only for this).**


In [ ]:
# Training accuracy
y_train_pred = log_reg_sklearn.predict(X_train_iteration_2)
training_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {training_accuracy}")
print(f"Test Accuracy: {accuracy}")

Training Accuracy: 0.8435374149659864
Test Accuracy: 0.8401360544217688


**a. What is the purpose of comparing those two values?**

*If the training accuracy is much higher than the test accuracy, it suggests that the model has overfit the training data.  This means the model has learned the training data's noise and specific details too well, rather than the underlying patterns.  As a result, it performs poorly on the test data, which is representative of data the model hasn't seen before.*

*Conversely, if the training and test accuracies are similar, it indicates that the model has generalized well and is likely to perform well on new data.  The goal in machine learning is to build models that generalize well, not just memorize the training data.*


**b. In this case, what does the comparison of those values suggest about the
model that you have built?**

*The difference between training accuracy (0.8458) and test accuracy (0.8452) is very small.
This suggests that the model generalizes well to unseen data.  There's not a significant
overfitting problem, as the model performs similarly on both the training and testing datasets.*

**P. Make up an employee. Assign this customer a value for each predictor variable
in this model, and store the results in a new dataframe. Now, put your employee
through this model.
a. What did your model predict -- will this employee stay or leave?**

In [ ]:
# Creating new employee DataFrame
new_employee = pd.DataFrame({
    'const': 1,
    'Age': [34],
    'DistanceFromHome': [27],
    'NumCompaniesWorked': [5],
    'JobSatisfaction_2': [0],           # JobSatisfaction: 1-Low; 2-Medium; 3-High; 4-Very High
    'JobSatisfaction_3': [1],
    'JobSatisfaction_4': [0],
    'MaritalStatus_Married': [1],
    'MaritalStatus_Single': [0],
    'WorkLifeBalance_2': [0],           # 1-Bad; 2-Good; 3-Better; 4-Best
    'WorkLifeBalance_3': [0],
    'WorkLifeBalance_4': [1],
    'EnvironmentSatisfaction_2': [0],   # 1-Low; 2-Medium; 3-High; 4-Very High
    'EnvironmentSatisfaction_3': [1],
    'EnvironmentSatisfaction_4': [0]
})

# Reordering columns to match the training data
new_employee = new_employee[X_train_iteration_2.columns]

# Prediction for the new employee
prediction = log_reg_sklearn.predict(new_employee)

# Interpreting the prediction (0 for stay, 1 for leave)
if prediction[0] == 0:
    print("The model predicts the employee will stay.")
else:
    print("The model predicts the employee will leave.")

The model predicts the employee will stay.


**b. According to your model, what is the probability that the employee will
leave?**

In [ ]:
probabilities = log_reg_sklearn.predict_proba(new_employee)
probability_of_leaving = probabilities[0][1]  # Probability of class 1 (leaving)

print(f"The probability that the employee will leave is: {probability_of_leaving:.4f}")

The probability that the employee will leave is: 0.2849


**Q. When using a logistic regression model to make predictions, why is it important
to only use values within the range of the dataset used to build the model?**

*Logistic regression models learn the relationship between predictor variables and the probability of a binary outcome based on the training data.  Extrapolating beyond the range of the training data can lead to unreliable predictions based on the following resons Non-linearity, Model Assumptions, Extrapolation Uncertainty, Overfitting and Underfitting.*

*The model's learned coefficients are optimized for the range of the data it was trained on.  Using values outside that range effectively forces the model to make assumptions that may not be justified, leading to potentially unreliable probability estimates. It's crucial to stay within the observed range of the predictors when making predictions with a logistic regression model to ensure the reliability of the results.*

**a. Make a new dataframe, but this time, for the numeric predictor variables,
select some numbers that are outside the range of the dataset -- do not
use a 400+ year-old vampire named “Mary.” Use your model to
make a prediction for this new dataframe. What do you notice about the
result? (To answer this, don’t simply state the predicted outcome, but
also write 1-2 sentences of explanation for what you see).**


In [ ]:
# Creating a new DataFrame with out-of-range values for numeric predictors
new_employee_out_of_range = pd.DataFrame({
    'const': 1,
    'Age': [85],                   # Max Age 60
    'DistanceFromHome': [50],       # Max DistanceFromHome 29
    'NumCompaniesWorked': [30],     # Max NumCompaniesWorked  09
    'JobSatisfaction_2': [0],       # JobSatisfaction: 1-Low; 2-Medium; 3-High; 4-Very High
    'JobSatisfaction_3': [1],
    'JobSatisfaction_4': [0],
    'MaritalStatus_Married': [1],
    'MaritalStatus_Single': [0],
    'WorkLifeBalance_2': [0],       # 1-Bad; 2-Good; 3-Better; 4-Best
    'WorkLifeBalance_3': [0],
    'WorkLifeBalance_4': [1],
    'EnvironmentSatisfaction_2': [0],   # 1-Low; 2-Medium; 3-High; 4-Very High
    'EnvironmentSatisfaction_3': [1],
    'EnvironmentSatisfaction_4': [0]
})

# Reorder columns to match the training data
new_employee_out_of_range = new_employee_out_of_range[X_train_iteration_2.columns]

# Prediction for the new employee with out-of-range values
prediction_out_of_range = log_reg_sklearn.predict(new_employee_out_of_range)
probability_out_of_range = log_reg_sklearn.predict_proba(new_employee_out_of_range)

print(f"Prediction for out-of-range employee: {prediction_out_of_range[0]}")

# Interpreting the prediction (0 for stay, 1 for leave)

if prediction_out_of_range[0] == 0:
    print("The model predicts the employee will stay.")
else:
    print("The model predicts the employee will leave.")

print(f"Probability of leaving for out-of-range employee: {probability_out_of_range[0,1]}")

Prediction for out-of-range employee: 0
The model predicts the employee will stay.
Probability of leaving for out-of-range employee: 0.3574215058149493


*The model predicts the employee will stay, even if this particular scenario is unrealistic (Age - 85, distance fromhome - 50 , NumCompaniesWorked - 30) . This is because the model is extrapolating beyond the data it was trained on, meaning that predictions will not reflect the real-world situation. The model has fit patterns within the range of the data it has observed, and cannot account for cases where the input features are very extreme.*

# **Part II: Random Forest Model**


**R. Read the dataset back into Python.**


In [ ]:
df1 = pd.read_csv(r'C:\Users\prabu\Downloads\IBM (1).csv')

In [ ]:
df1.head()

Age Attrition              Department  DistanceFromHome  Education  \
0   41       Yes                   Sales                 1          2   
1   49        No  Research & Development                 8          1   
2   37       Yes  Research & Development                 2          2   
3   33        No  Research & Development                 3          4   
4   27        No  Research & Development                 2          1   

  EducationField  EnvironmentSatisfaction  JobSatisfaction MaritalStatus  \
0  Life Sciences                        2                4        Single   
1  Life Sciences                        3                2       Married   
2          Other                        4                3        Single   
3  Life Sciences                        4                3       Married   
4        Medical                        1                2       Married   

   MonthlyIncome  NumCompaniesWorked  WorkLifeBalance  YearsAtCompany  
0           5993                   8                1               6  
1           5130                   1                3              10  
2           2090                   6                3               0  
3           2909                   1                3               8  
4           3468                   9                3               2

**S. Dummify the categorical inputs again, but this time, don’t drop any levels. Again,
use the label encoder to convert ‘Attrition’ into a binary format.**


In [ ]:
# Converting 'Attrition' to binary format using LabelEncoder

le = LabelEncoder()
df1['Attrition'] = le.fit_transform(df1['Attrition'])

# Dummifying categorical features without dropping any levels

nominal_columns = ['Department', 'EducationField', 'MaritalStatus']
ordinal_columns = {
    'Education': [1, 2, 3, 4, 5],
    'EnvironmentSatisfaction': [1, 2, 3, 4],
    'JobSatisfaction': [1, 2, 3, 4],
    'WorkLifeBalance': [1, 2, 3, 4]
}

# Converting ordinal variables to ordered categorical
for col, order in ordinal_columns.items():
    df1[col] = pd.Categorical(df1[col], categories=order, ordered=True)

# using pd.get_dummies without drop_first=True
df1_encoded = pd.get_dummies(df1, columns=nominal_columns + list(ordinal_columns.keys()), drop_first=False)
df1_encoded = df1_encoded.astype(int)
df1_encoded.head()

Age  Attrition  DistanceFromHome  MonthlyIncome  NumCompaniesWorked  \
0   41          1                 1           5993                   8   
1   49          0                 8           5130                   1   
2   37          1                 2           2090                   6   
3   33          0                 3           2909                   1   
4   27          0                 2           3468                   9   

   YearsAtCompany  Department_Human Resources  \
0               6                           0   
1              10                           0   
2               0                           0   
3               8                           0   
4               2                           0   

   Department_Research & Development  Department_Sales  \
0                                  0                 1   
1                                  1                 0   
2                                  1                 0   
3                                  1                 0   
4                                  1                 0   

   EducationField_Human Resources  ...  EnvironmentSatisfaction_3  \
0                               0  ...                          0   
1                               0  ...                          1   
2                               0  ...                          0   
3                               0  ...                          0   
4                               0  ...                          0   

   EnvironmentSatisfaction_4  JobSatisfaction_1  JobSatisfaction_2  \
0                          0                  0                  0   
1                          0                  0                  1   
2                          1                  0                  0   
3                          1                  0                  0   
4                          0                  0                  1   

   JobSatisfaction_3  JobSatisfaction_4  WorkLifeBalance_1  WorkLifeBalance_2  \
0                  0                  1                  1                  0   
1                  0                  0                  0                  0   
2                  1                  0                  0                  0   
3                  1                  0                  0                  0   
4                  0                  0                  0                  0   

   WorkLifeBalance_3  WorkLifeBalance_4  
0                  0                  0  
1                  1                  0  
2                  1                  0  
3                  1                  0  
4                  1                  0  

[5 rows x 35 columns]

**T. Re-partition the data, using the same seed value that you used in the previous part of
this assignment.**

In [ ]:
X = df1_encoded.drop('Attrition', axis=1)
y = df1_encoded['Attrition']

# Splitting data into training and testing sets, using the same random_state as before
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=911)

# Displaying the sizes of the training and test sets
train_size = X_train.shape[0]
test_size = X_test.shape[0]

print(train_size, test_size)
print(metrics.classification_report(y_test, y_pred))

882 588
              precision    recall  f1-score   support

           0       0.84      1.00      0.91       490
           1       0.75      0.06      0.11        98

    accuracy                           0.84       588
   macro avg       0.80      0.53      0.51       588
weighted avg       0.83      0.84      0.78       588



**U. Build a random forest model in Python with your training set. Use the same input
variables, and same output variable, as you used in the first logistic regression model
(the only difference here is that the categories should not have any levels dropped).
Use GridSearchCV to help you determine the best hyperparameter settings
for your mode**

In [ ]:
# Defining the parameter grid for hyperparameter tuning
param_grid = {
    'n_estimators': [50, 100, 150],
    'max_depth': [2, 4, 6, 8],
    'max_features': [4, 8, 16, 20],
    'min_samples_leaf': [ 2, 4 , 6, 10 ]
}

# Initializing the RandomForestClassifier
rf_classifier = RandomForestClassifier(random_state=911)

# GridSearchCV to find the best hyperparameters
grid_search = GridSearchCV(estimator=rf_classifier, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)

# Best model and its hyperparameters
best_rf_model = grid_search.best_estimator_
best_params = grid_search.best_params_

print("Best Hyperparameters:", best_params)

# Evaluate the best model on the test set
y_pred = best_rf_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Test Accuracy:", accuracy)

Best Hyperparameters: {'max_depth': 8, 'max_features': 8, 'min_samples_leaf': 2, 'n_estimators': 100}
Test Accuracy: 0.8367346938775511


**V. How did your random forest model rank the variables in order of importance, from
highest to lowest? For a random forest model, how can you interpret feature
importance?**


In [ ]:
# Getting feature importances from the trained RandomForestClassifier
feature_importances = best_rf_model.feature_importances_

# Creating a DataFrame to display feature importances
feature_importance_df = pd.DataFrame({'Feature': X_train.columns, 'Importance': feature_importances})

# Sorting the DataFrame by importance in descending order
feature_importance_df = feature_importance_df.sort_values(by='Importance', ascending=False)

# Feature importances
feature_importance_df

Feature  Importance
2                       MonthlyIncome    0.209612
0                                 Age    0.138412
4                      YearsAtCompany    0.124194
1                    DistanceFromHome    0.095744
3                  NumCompaniesWorked    0.052369
16               MaritalStatus_Single    0.031987
29                  JobSatisfaction_4    0.020472
32                  WorkLifeBalance_3    0.017574
11             EducationField_Medical    0.016868
18                        Education_2    0.016693
25          EnvironmentSatisfaction_4    0.016542
31                  WorkLifeBalance_2    0.015846
20                        Education_4    0.015683
22          EnvironmentSatisfaction_1    0.015627
9        EducationField_Life Sciences    0.015549
28                  JobSatisfaction_3    0.015253
6   Department_Research & Development    0.015022
7                    Department_Sales    0.014240
13    EducationField_Technical Degree    0.013697
19                        Education_3    0.013299
15              MaritalStatus_Married    0.013150
10           EducationField_Marketing    0.012621
30                  WorkLifeBalance_1    0.012606
14             MaritalStatus_Divorced    0.011976
26                  JobSatisfaction_1    0.011704
24          EnvironmentSatisfaction_3    0.010902
23          EnvironmentSatisfaction_2    0.010901
17                        Education_1    0.010554
27                  JobSatisfaction_2    0.010179
33                  WorkLifeBalance_4    0.008325
5          Department_Human Resources    0.005032
12               EducationField_Other    0.003582
8      EducationField_Human Resources    0.003489
21                        Education_5    0.000295

*The random forest model ranked features by their importance in predicting Attrition, with MonthlyIncome, Age, and YearsAtCompany being the top three. Feature importance in a random forest indicates how much each variable contributes to reducing prediction error across the trees. Higher importance means the feature is more useful for distinguishing between classes, though it doesn’t imply causation just predictive power. This information helps prioritize features and guide feature selection for simplified models.*

**W. Assess the performance of your model against the test set. Build a confusion matrix to
do this. You can use Python functions to answer any of these questions or you can use
your confusion matrix to determine the answers in a slightly more manual way. The
‘positive’ class in this model is represented by the “1” outcome.
a. What is your model’s accuracy rate?
b. What is your model’s sensitivity rate?
c. What is your model’s specificity rate?
d. What is your model’s precision?
e. What is your model’s balanced accuracy?**

In [ ]:
# Confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
sensitivity = recall_score(y_test, y_pred)
specificity = cm[0, 0] / (cm[0, 0] + cm[0, 1])
precision = precision_score(y_test, y_pred)
balanced_accuracy = (sensitivity + specificity) / 2


print("Confusion Matrix:\n", cm)
print(f"\nAccuracy: {accuracy}")
print(f"Sensitivity: {sensitivity}")
print(f"Specificity: {specificity}")
print(f"Precision: {precision}")
print(f"Balanced Accuracy: {balanced_accuracy}")

Confusion Matrix:
 [[482   8]
 [ 88  10]]

Accuracy: 0.8367346938775511
Sensitivity: 0.10204081632653061
Specificity: 0.9836734693877551
Precision: 0.5555555555555556
Balanced Accuracy: 0.5428571428571429


**X. Compare your model’s accuracy against the training set vs. your model’s accuracy against
the test set. How different were these results?**

In [ ]:
# Training accuracy
y_train_pred = best_rf_model.predict(X_train)
training_accuracy = accuracy_score(y_train, y_train_pred)
print(f"Training Accuracy: {training_accuracy}")

# Test accuracy (already calculated in a previous code block)
print(f"Test Accuracy: {accuracy}")

Training Accuracy: 0.9104308390022676
Test Accuracy: 0.8367346938775511


**Y. Use the predict() function with your model to classify the person who you invented in
the previous section. Does the model think that the employee will leave?**

In [ ]:
# Creating a new DataFrame with out-of-range values for predictors
new_employee_out_of_range = pd.DataFrame({
    'Age': [85],                    # Max Age 60
    'DistanceFromHome': [50],       # Max DistanceFromHome 29
    'NumCompaniesWorked': [30],     # Max NumCompaniesWorked  09
    'Department_Human Resources': [0],
    'Department_Research & Development': [0],
    'Department_Sales': [0],
    'EducationField_Human Resources': [0],
    'EducationField_Life Sciences': [0],
    'EducationField_Marketing': [0],
    'EducationField_Medical': [0],
    'EducationField_Other': [0],
    'EducationField_Technical Degree': [0],
    'MaritalStatus_Divorced': [0],
    'MaritalStatus_Married': [1],
    'MaritalStatus_Single': [0],
    'Education_1': [0],
    'Education_2': [0],
    'Education_3': [0],
    'Education_4': [0],
    'Education_5': [0],
    'EnvironmentSatisfaction_1': [0],
    'EnvironmentSatisfaction_2': [0],
    'EnvironmentSatisfaction_3': [1],
    'EnvironmentSatisfaction_4': [0],
    'JobSatisfaction_1': [0],
    'JobSatisfaction_2': [0],
    'JobSatisfaction_3': [1],
    'JobSatisfaction_4': [0],
    'WorkLifeBalance_1': [0],
    'WorkLifeBalance_2': [0],
    'WorkLifeBalance_3': [0],
    'WorkLifeBalance_4': [1],
    'MonthlyIncome': [0],
    'YearsAtCompany': [0]
})

# Reordering columns to match the training data
new_employee_out_of_range = new_employee_out_of_range[X_train.columns]

# Prediction for the new employee with out-of-range values
prediction_out_of_range = best_rf_model.predict(new_employee_out_of_range)
probability_out_of_range = best_rf_model.predict_proba(new_employee_out_of_range)

print(f"Prediction for out-of-range employee by Random Forest model: {prediction_out_of_range[0]}")

# Interpreting the prediction (0 for stay, 1 for leave)
if prediction_out_of_range[0] == 0:
    print("The Random Forest model predicts the employee will stay.")
else:
    print("The Random Forest model predicts the employee will leave.")

print(f"Probability of leaving for out-of-range employee by Random Forest Model : {probability_out_of_range[0,1]}")

Prediction for out-of-range employee by Random Forest model: 0
The Random Forest model predicts the employee will stay.
Probability of leaving for out-of-range employee by Random Forest Model : 0.4407797619047619


**Z.
Write a 3-5 sentence paragraph that speculates about how a company’s Human
Resources Department might be able to use the models you have built for this
assignment**

*Human Resources (HR) department can use this models developed in this assignment, since it provides a data-driven approach to understand factors influencing employee attrition.*

*The logistic regression model indicates that Age (with older employees being less likely to leave), DistanceFromHome (employees living farther away are more likely to leave), and JobSatisfaction_4 (very high job satisfaction reduces attrition risk) are significant predictors.*

*The random forest model further highlights MonthlyIncome, YearsAtCompany, and MaritalStatus_Single as critical features, with single employees and those with lower incomes being at higher risk of leaving.*

*By leveraging these insights, HR can design targeted interventions, such as salary adjustments for low-income employees, career development opportunities for newer employees, and personalized support for those with long commutes.*

*Additionally, HR can monitor employees with high attrition probabilities to proactively address their concerns and improve overall retention.*